// docusaurus-plugin-typedoc@1.0.0-next.2/dist/theme.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocusaurusTheme = void 0;
const fs = __importStar(require("fs"));
const typedoc_1 = require("typedoc");
const typedoc_plugin_markdown_1 = require("typedoc-plugin-markdown");
const navigation_1 = require("./navigation");
const theme_context_1 = require("./theme-context");
class DocusaurusTheme extends typedoc_plugin_markdown_1.MarkdownTheme {
    constructor(renderer) {
        super(renderer);
        this.writeCategoryYaml = (categoryPath, label, position, collapsed) => {
            const yaml = [`label: "${label}"`];
            if (position !== null) {
                yaml.push(`position: ${position}`);
            }
            if (!collapsed) {
                yaml.push(`collapsed: false`);
            }
            if (fs.existsSync(categoryPath)) {
                fs.writeFileSync(categoryPath + '/_category_.yml', yaml.join('\n'));
            }
        };
        this.listenTo(this.application.renderer, {
            [typedoc_1.PageEvent.END]: this.onPageEnd,
            [typedoc_1.RendererEvent.END]: this.onRendererEnd,
        });
    }
    getRenderContext() {
        this._contextCache || (this._contextCache = new theme_context_1.DocusaurusThemeRenderContext(this, this.application.options));
        return this._contextCache;
    }
    onPageEnd(page) {
        if (page.contents) {
            page.contents = page.contents.replace(/\\</g, '<');
        }
    }
    onRendererEnd(renderer) {
        if (this.sidebar.autoConfiguration) {
            this.writeCategoryYaml(renderer.outputDirectory, this.sidebar.categoryLabel, this.sidebar.position, this.sidebar.collapsed);
            this.loopAndWriteCategories(renderer.outputDirectory);
        }
    }
    loopAndWriteCategories(path) {
        const directory = fs.readdirSync(path);
        directory.forEach((segment) => {
            const fullPath = `${path}/${segment}`;
            const isDirectory = fs.lstatSync(fullPath).isDirectory();
            if (isDirectory) {
                const mapping = Object.entries(this.mappings)
                    .filter((entry) => {
                    return entry[1].directory === segment;
                })
                    .map((entry) => entry[1])[0];
                const subdirectory = fs.readdirSync(fullPath);
                const containsDir = subdirectory.some((item) => fs.lstatSync(`${fullPath}/${item}`).isDirectory());
                if (mapping && !containsDir) {
                    this.writeCategoryYaml(fullPath, (0, navigation_1.getKindPlural)(mapping.kind), navigation_1.CATEGORY_POSITION[mapping.kind], true);
                }
                this.loopAndWriteCategories(fullPath);
            }
        });
    }
}
__decorate([
    (0, typedoc_1.BindOption)('sidebar')
], DocusaurusTheme.prototype, "sidebar", void 0);
exports.DocusaurusTheme = DocusaurusTheme;
