// typedoc-plugin-markdown@4.0.0-next.2/dist/support/els.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.indentBlock = exports.table = exports.codeBlock = exports.horizontalRule = exports.unorderedList = exports.backTicks = exports.italic = exports.bold = exports.link = exports.heading = void 0;
const utils_1 = require("./utils");
const heading = (level, text) => `${[...Array(level)].map(() => '#').join('')} ${text}`;
exports.heading = heading;
const link = (label, url) => url ? `[${label}](${url})` : '';
exports.link = link;
const bold = (text) => `**${text}**`;
exports.bold = bold;
const italic = (text) => `*${text}*`;
exports.italic = italic;
const backTicks = (text) => `\`${text}\``;
exports.backTicks = backTicks;
const unorderedList = (items) => items.map((item) => `- ${item}`).join('\n');
exports.unorderedList = unorderedList;
const horizontalRule = () => '\n\n---';
exports.horizontalRule = horizontalRule;
const codeBlock = (content) => '```ts\n' + (0, utils_1.unEscapeChars)(content) + '\n```';
exports.codeBlock = codeBlock;
const table = (headers, rows) => `\n| ${headers.join(' | ')} |\n| ${headers
    .map(() => ':------')
    .join(' | ')} |\n${rows.map((row) => `| ${row.join(' | ')} \n`).join('')}`;
exports.table = table;
const indentBlock = (content) => {
    const lines = (0, utils_1.formatContents)(content).split('\n');
    return lines.map((line) => (line.length ? `> ${line}` : '>')).join('\n');
};
exports.indentBlock = indentBlock;
