// typedoc-plugin-markdown@4.0.0-next.2/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkdownThemeRenderContext = exports.MarkdownTheme = exports.partials = exports.load = void 0;
const typedoc_1 = require("typedoc");
const theme_1 = require("./theme");
Object.defineProperty(exports, "MarkdownTheme", { enumerable: true, get: function () { return theme_1.MarkdownTheme; } });
const theme_context_1 = require("./theme-context");
Object.defineProperty(exports, "MarkdownThemeRenderContext", { enumerable: true, get: function () { return theme_context_1.MarkdownThemeRenderContext; } });
function load(app) {
    app.renderer.defineTheme('markdown', theme_1.MarkdownTheme);
    app.options.addReader(new (class {
        constructor() {
            this.priority = 1000;
            this.name = 'markdown-theme-reader';
        }
        read(container) {
            if (container.getValue('theme') === 'default') {
                container.setValue('theme', 'markdown');
            }
        }
    })());
    app.options.addDeclaration({
        name: 'entryDocument',
        help: '[Markdown Plugin] The file name of the entry document.',
        type: typedoc_1.ParameterType.String,
        defaultValue: 'README.md',
    });
    app.options.addDeclaration({
        name: 'flattenOutput',
        help: '[Markdown Plugin] Flatten output files.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'symbolsWithOwnFile',
        help: "[Markdown Plugin] Specifies which symbols are contained in their own file. Values 'none', 'all' OR Array of ['class', 'interface', 'enum', 'function', 'variable', 'type']",
        type: typedoc_1.ParameterType.String | typedoc_1.ParameterType.Array,
        defaultValue: 'all',
    });
    app.options.addDeclaration({
        name: 'hidePageTitle',
        help: '[Markdown Plugin] Do not print page title.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'hideBreadcrumbs',
        help: '[Markdown Plugin] Do not print breadcrumbs.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'hideInPageTOC',
        help: '[Markdown Plugin] Do not render in-page table of contents items.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'hideHierarchy',
        help: '[Markdown Plugin] Do not print reflection hierarchy.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'longTitles',
        help: '[Markdown Plugin] Display full name including module paths in page titles.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'typeDeclarationStyle',
        help: '[Markdown Plugin] Specify the render style of type declarations.',
        type: typedoc_1.ParameterType.String,
        defaultValue: 'list',
        validate: (option) => {
            const availableValues = ['table', 'list'];
            if (!availableValues.includes(option)) {
                throw new Error(`Unexpected value for typeDeclarationStyle, the expected value is one of 'table', 'list'`);
            }
        },
    });
    app.options.addDeclaration({
        name: 'enableFrontmatter',
        help: '[Markdown Plugin] Prepend frontmatter to output.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'frontmatterTags',
        help: '[Markdown Plugin] Specify which file comment tags should be added to frontmatter.',
        type: typedoc_1.ParameterType.Array,
        defaultValue: [],
    });
    app.options.addDeclaration({
        name: 'frontmatterGlobals',
        help: '[Markdown Plugin] Specify static variables to be added to all frontmatter.',
        type: typedoc_1.ParameterType.Object,
        defaultValue: {},
    });
    app.options.addDeclaration({
        help: '[Markdown Plugin] Specifies the base url for internal link. If omitted all urls will be relative.',
        name: 'baseUrl',
        type: typedoc_1.ParameterType.String,
    });
    app.options.addDeclaration({
        name: 'namedAnchors',
        help: '[Markdown Plugin] Use HTML named anchors as fragment identifiers for engines that do not automatically assign header ids. Should be set for Bitbucket Server docs.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
    app.options.addDeclaration({
        name: 'preserveAnchorCasing',
        help: '[Markdown Plugin] Preserve anchor casing when generating links.',
        type: typedoc_1.ParameterType.Boolean,
        defaultValue: false,
    });
}
exports.load = load;
__exportStar(require("./models"), exports);
__exportStar(require("./options-reader"), exports);
var resources_1 = require("./resources");
Object.defineProperty(exports, "partials", { enumerable: true, get: function () { return resources_1.partials; } });
