// typedoc-plugin-markdown@4.0.0-next.2/dist/theme-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkdownThemeRenderContext = void 0;
const path = require("path");
const resources_1 = require("./resources");
const constants_1 = require("./support/constants");
const helpers_1 = require("./support/helpers");
const utils_1 = require("./support/utils");
class MarkdownThemeRenderContext {
    set activeLocation(activeLocation) {
        this._activeLocation = activeLocation;
    }
    get activeLocation() {
        return this._activeLocation;
    }
    constructor(theme, options) {
        this.theme = theme;
        this.options = options;
        this.globalsFile = 'modules.md';
        this.urlTo = (reflection) => this.relativeURL(reflection.url);
        this.relativeURL = (url) => {
            return this.getRelativeUrl(url);
        };
        this.baseFrontmatterVars = (page) => {
            return this.getBaseFrontmatterVars(page);
        };
        this.templates = (0, resources_1.templates)(this);
        this.partials = (0, resources_1.partials)(this);
    }
    getOption(name) {
        return this.options.getValue(name);
    }
    getRelativeUrl(url) {
        if (!url) {
            return null;
        }
        if (constants_1.URL_PREFIX.test(url)) {
            return url;
        }
        else {
            if (this.getOption('baseUrl')) {
                return this.getOption('baseUrl') + url.replace(/\\/g, '/');
            }
            const relative = path.relative(path.dirname(this.activeLocation), path.dirname(url));
            return path.join(relative, path.basename(url)).replace(/\\/g, '/');
        }
    }
    getBaseFrontmatterVars(page) {
        return {
            title: (0, utils_1.unEscapeChars)((0, helpers_1.getReflectionTitle)(page.model)),
        };
    }
}
exports.MarkdownThemeRenderContext = MarkdownThemeRenderContext;
